File: Components/DevExpressVCL/Demos/ExpressQuantumGrid/CBuilder/ColumnsSimpleDemo/ColumnsSimpleDemoMain.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#include "shellapi.hpp"
#pragma hdrstop

#include "ColumnsSimpleDemoMain.h"
#include "AboutDemoForm.h"
#include "ColumnsSimpleDemoData.h"
#include "ColumnsSimpleDemoCars.h"
#include "ColumnsSimpleDemoCities.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "cxClasses"
#pragma link "cxControls"
#pragma link "cxCustomData"
#pragma link "cxData"
#pragma link "cxDBData"
#pragma link "cxEdit"
#pragma link "cxFilter"
#pragma link "cxGraphics"
#pragma link "cxGrid"
#pragma link "cxGridCustomTableView"
#pragma link "cxGridCustomView"
#pragma link "cxGridDBTableView"
#pragma link "cxGridLevel"
#pragma link "cxGridTableView"
#pragma link "cxStyles"
#pragma link "cxLookAndFeels"
#pragma link "cxBlobEdit"
#pragma link "cxButtonEdit"
#pragma link "cxCalc"
#pragma link "cxCheckBox"
#pragma link "cxDataStorage"
#pragma link "cxDBLookupComboBox"
#pragma link "cxDropDownEdit"
#pragma link "cxImageComboBox"
#pragma link "cxMaskEdit"
#pragma link "cxMRUEdit"
#pragma link "cxSpinEdit"
#pragma link "cxTimeEdit"
#pragma link "cxGridCardView"
#pragma link "cxLookAndFeelPainters"
#pragma resource "*.dfm"
TColumnsSimpleDemoMainForm *ColumnsSimpleDemoMainForm;
//---------------------------------------------------------------------------
__fastcall TColumnsSimpleDemoMainForm::TColumnsSimpleDemoMainForm(TComponent* Owner)
  : TfmBaseForm(Owner)
{
}
//---------------------------------------------------------------------------

TcxGridTableView* TColumnsSimpleDemoMainForm::GetView()
{
  return (TcxGridTableView*)Grid->FocusedView;
}
//---------------------------------------------------------------------------

void __fastcall TColumnsSimpleDemoMainForm::UpdateMenu()
{
  miShowEditButtons->Items[(int)View->OptionsView->ShowEditButtons]->Checked = true;

  miCellSelect->Checked = View->OptionsSelection->CellSelect;
  miHideFocusRect->Checked = View->OptionsSelection->HideFocusRect;
  miHideSelection->Checked = View->OptionsSelection->HideSelection;
  miInvertSelect->Checked = View->OptionsSelection->InvertSelect;
  miMultiSelect->Checked = View->OptionsSelection->MultiSelect;
}
//---------------------------------------------------------------------------

void __fastcall TColumnsSimpleDemoMainForm::FormShow(TObject *Sender)
{
  if(tvCustomers->Controller->FocusedRecord != NULL)
    tvCustomers->Controller->FocusedRecord->Expanded = true;
}
//---------------------------------------------------------------------------

void __fastcall TColumnsSimpleDemoMainForm::miEditButtonsClick(TObject *Sender)
{
  View->OptionsView->ShowEditButtons = (TcxGridShowEditButtons)((TMenuItem*)Sender)->MenuIndex;
  UpdateMenu();
}
//---------------------------------------------------------------------------

void __fastcall TColumnsSimpleDemoMainForm::miCellSelectClick(TObject *Sender)
{
  View->OptionsSelection->CellSelect = !View->OptionsSelection->CellSelect;
  UpdateMenu();
}
//---------------------------------------------------------------------------

void __fastcall TColumnsSimpleDemoMainForm::miHideFocusRectClick(TObject *Sender)
{
  View->OptionsSelection->HideFocusRectOnExit = !View->OptionsSelection->HideFocusRectOnExit;
  UpdateMenu();
}
//---------------------------------------------------------------------------

void __fastcall TColumnsSimpleDemoMainForm::miHideSelectionClick(TObject *Sender)
{
  View->OptionsSelection->HideSelection = !View->OptionsSelection->HideSelection;
  UpdateMenu();
}
//---------------------------------------------------------------------------

void __fastcall TColumnsSimpleDemoMainForm::miInvertSelectClick(TObject *Sender)
{
  View->OptionsSelection->InvertSelect = !View->OptionsSelection->InvertSelect;
  UpdateMenu();
}
//---------------------------------------------------------------------------

void __fastcall TColumnsSimpleDemoMainForm::miMultiSelectClick(TObject *Sender)
{
  View->OptionsSelection->MultiSelect = !View->OptionsSelection->MultiSelect;
  UpdateMenu();
}
//---------------------------------------------------------------------------

void __fastcall TColumnsSimpleDemoMainForm::tvCustomersStatePropertiesButtonClick(TObject *Sender)
{
  MessageDlg("Button click event handler", mtInformation, TMsgDlgButtons() << mbOK, 0);
}
//---------------------------------------------------------------------------

void __fastcall TColumnsSimpleDemoMainForm::tvCustomersCityPropertiesButtonClick(
      TObject *Sender, int AButtonIndex)
{
  switch(AButtonIndex) {
    case 0:
      if (ColumnsSimpleDemoCitiesForm->ShowModal() == mrOk) {
        ColumnsSimpleDemoDataDM->tblCustomers->Edit();
        ColumnsSimpleDemoDataDM->tblCustomersCity->Value = ColumnsSimpleDemoCitiesForm->Value;
      }; break;
    case 1:
      ShellExecute(Handle, "OPEN", "http://www.usacitiesonline.com/", NULL, NULL, SW_SHOWMAXIMIZED); break;
  }
}
//---------------------------------------------------------------------------

void __fastcall TColumnsSimpleDemoMainForm::tvOrdersCarInfoPropertiesInitPopup(TObject *Sender)
{
  ColumnSimpleDemoCarsForm->bvCars->DataController->LocateByKey(
    ColumnsSimpleDemoDataDM->tblOrdersProductID->AsInteger);
}
//---------------------------------------------------------------------------

void __fastcall TColumnsSimpleDemoMainForm::tvOrdersCarInfoPropertiesCloseUp(TObject *Sender)
{
  ColumnsSimpleDemoDataDM->tblOrders->Edit();
  ColumnsSimpleDemoDataDM->tblOrdersProductID->Value = dmGridCars->mdModelsID->Value;
}
//---------------------------------------------------------------------------

void __fastcall TColumnsSimpleDemoMainForm::GridFocusedViewChanged(
      TcxCustomGrid *Sender, TcxCustomGridView *APrevFocusedView,
      TcxCustomGridView *AFocusedView)
{
  UpdateMenu();
}
//---------------------------------------------------------------------------

void __fastcall TColumnsSimpleDemoMainForm::tvCustomersCompanyGetCellHint(
      TcxCustomGridTableItem *Sender, TcxCustomGridRecord *ARecord,
      TcxGridTableDataCellViewInfo *ACellViewInfo, const TPoint &AMousePos,
      TCaption &AHintText, bool &AIsHintMultiLine, TRect &AHintTextRect)
{
  AHintText = ARecord->DisplayTexts[tvCustomersAddress->Index] + CHAR(13) +
    ARecord->DisplayTexts[tvCustomersCity->Index] + ", " + ARecord->DisplayTexts[tvCustomersState->Index] + " " +
    ARecord->DisplayTexts[tvCustomersZipCode->Index] + CHAR(13) +
    ARecord->DisplayTexts[tvCustomersFaxPhone->Index];
  AHintTextRect.Top = ACellViewInfo->Bounds.Bottom + 3;
}
//---------------------------------------------------------------------------

File: Components/DevExpressVCL/Demos/ExpressQuantumGrid/Delphi/ColumnsSimpleDemo/ColumnsSimpleDemoMain.pas
unit ColumnsSimpleDemoMain;

{$I cxVer.inc}

interface

uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms,
  Dialogs, StdCtrls, cxGridLevel, cxGridCustomTableView, cxGridTableView,
  cxGridDBTableView, cxControls, cxGridCustomView, cxGrid, cxCustomData,
  ExtCtrls, ActnList, ImgList, Menus, ComCtrls, cxGridCustomPopupMenu,
  cxGridPopupMenu, Grids, DBGrids,
  cxDBData, cxStyles, cxGraphics, cxFilter, cxData, cxEdit, DB, cxClasses,
  cxDataStorage, cxMRUEdit, cxButtonEdit, cxDropDownEdit, cxCheckBox,
  cxMaskEdit, cxDBLookupComboBox, cxTimeEdit, cxImageComboBox, cxSpinEdit,
  cxCalc, cxBlobEdit, cxImage, cxRadioGroup, cxMemo, cxHyperLinkEdit,
  cxLookAndFeels, cxLookAndFeelPainters, BaseForm, cxGridCardView, CarsDataForGrid, cxNavigator;

type
  TColumnsSimpleDemoMainForm = class(TfmBaseForm)
    miOptions: TMenuItem;
    Grid: TcxGrid;
    lvCustomers: TcxGridLevel;
    lvOrders: TcxGridLevel;
    tvCustomers: TcxGridDBTableView;
    tvCustomersFirstName: TcxGridDBColumn;
    tvCustomersLastName: TcxGridDBColumn;
    tvCustomersCompany: TcxGridDBColumn;
    tvCustomersPrefix: TcxGridDBColumn;
    tvCustomersAddress: TcxGridDBColumn;
    tvCustomersCity: TcxGridDBColumn;
    tvCustomersState: TcxGridDBColumn;
    tvCustomersZipCode: TcxGridDBColumn;
    tvCustomersCustomer: TcxGridDBColumn;
    tvCustomersFaxPhone: TcxGridDBColumn;
    tvOrders: TcxGridDBTableView;
    tvOrdersProductID: TcxGridDBColumn;
    tvOrdersPurchaseDate: TcxGridDBColumn;
    tvOrdersTime: TcxGridDBColumn;
    tvOrdersPaymentType: TcxGridDBColumn;
    tvOrdersDescription: TcxGridDBColumn;
    tvOrdersQuantity: TcxGridDBColumn;
    tvOrdersPaymentAmount: TcxGridDBColumn;
    tvOrdersCarInfo: TcxGridDBColumn;
    miShowEditButtons: TMenuItem;
    miEditButtonsNever: TMenuItem;
    miEditButtonsForFocusedRecord: TMenuItem;
    miEditButtonsAlways: TMenuItem;
    miOptionsSelection: TMenuItem;
    miCellSelect: TMenuItem;
    miHideFocusRect: TMenuItem;
    miHideSelection: TMenuItem;
    miInvertSelect: TMenuItem;
    miMultiSelect: TMenuItem;
    procedure FormShow(Sender: TObject);
    procedure tvCustomersStatePropertiesButtonClick(Sender: TObject);
    procedure tvCustomersCityPropertiesButtonClick(Sender: TObject;
      AButtonIndex: Integer);
    procedure tvOrdersCarInfoPropertiesInitPopup(Sender: TObject);
    procedure tvOrdersCarInfoPropertiesCloseUp(Sender: TObject);
    procedure GridFocusedViewChanged(Sender: TcxCustomGrid;
      APrevFocusedView, AFocusedView: TcxCustomGridView);
    procedure miEditButtonsClick(Sender: TObject);
    procedure miCellSelectClick(Sender: TObject);
    procedure miHideFocusRectClick(Sender: TObject);
    procedure miHideSelectionClick(Sender: TObject);
    procedure miInvertSelectClick(Sender: TObject);
    procedure miMultiSelectClick(Sender: TObject);
    procedure tvCustomersCompanyGetCellHint(Sender: TcxCustomGridTableItem;
      ARecord: TcxCustomGridRecord; ACellViewInfo: TcxGridTableDataCellViewInfo;
      const AMousePos: TPoint; var AHintText: TCaption;
      var AIsHintMultiLine: Boolean; var AHintTextRect: TRect);
  private
    function GetView: TcxGridTableView;
    procedure UpdateMenu;
  public
    property View: TcxGridTableView read GetView;
  end;

var
  ColumnsSimpleDemoMainForm: TColumnsSimpleDemoMainForm;

implementation

uses
  ShellAPI, ColumnsSimpleDemoData, ColumnsSimpleDemoCities,
  ColumnsSimpleDemoCars, AboutDemoForm;

{$R *.dfm}

function TColumnsSimpleDemoMainForm.GetView: TcxGridTableView;
begin
  Result := TcxGridTableView(Grid.FocusedView);
end;

procedure TColumnsSimpleDemoMainForm.UpdateMenu;
begin
  MenuItemCheckSubItemWithTag('miShowEditButtons', Ord(View.OptionsView.ShowEditButtons));
  MenuItemSetChecked('miCellSelect', View.OptionsSelection.CellSelect);
  MenuItemSetChecked('miHideFocusRect', View.OptionsSelection.HideFocusRect);
  MenuItemSetChecked('miHideSelection', View.OptionsSelection.HideSelection);
  MenuItemSetChecked('miInvertSelect', View.OptionsSelection.InvertSelect);
  MenuItemSetChecked('miMultiSelect', View.OptionsSelection.MultiSelect);
end;

procedure TColumnsSimpleDemoMainForm.FormShow(Sender: TObject);
begin
  if tvCustomers.Controller.FocusedRecord <> nil then
    tvCustomers.Controller.FocusedRecord.Expanded := True;
  UpdateMenu;
end;

procedure TColumnsSimpleDemoMainForm.tvCustomersStatePropertiesButtonClick(Sender: TObject);
begin
  MessageDlg('Button click event handler', mtInformation, [mbOK], 0);
end;

procedure TColumnsSimpleDemoMainForm.tvCustomersCityPropertiesButtonClick(
  Sender: TObject; AButtonIndex: Integer);
begin
  case AButtonIndex of
    0: if ColumnsSimpleDemoCitiesForm.ShowModal = mrOk then
       begin
         ColumnsSimpleDemoDataDM.tblCustomers.Edit;
         ColumnsSimpleDemoDataDM.tblCustomersCity.Value := ColumnsSimpleDemoCitiesForm.Value;
       end;
    1:
       ShellExecute(Handle, PChar('OPEN'), PChar('http://www.usacitiesonline.com/'), nil, nil, SW_SHOWMAXIMIZED);
  end;
end;

procedure TColumnsSimpleDemoMainForm.tvCustomersCompanyGetCellHint(
  Sender: TcxCustomGridTableItem; ARecord: TcxCustomGridRecord;
  ACellViewInfo: TcxGridTableDataCellViewInfo; const AMousePos: TPoint;
  var AHintText: TCaption; var AIsHintMultiLine: Boolean;
  var AHintTextRect: TRect);
begin
  AHintText := ARecord.DisplayTexts[tvCustomersAddress.Index] + #13 +
    ARecord.DisplayTexts[tvCustomersCity.Index] + ', ' + ARecord.DisplayTexts[tvCustomersState.Index] + ' ' +
    ARecord.DisplayTexts[tvCustomersZipCode.Index] + #13 +
    ARecord.DisplayTexts[tvCustomersFaxPhone.Index];
  AHintTextRect.Top := ACellViewInfo.Bounds.Bottom + 3;  
end;

procedure TColumnsSimpleDemoMainForm.tvOrdersCarInfoPropertiesInitPopup(Sender: TObject);
begin
  ColumnSimpleDemoCarsForm.bvCars.DataController.LocateByKey(
    ColumnsSimpleDemoDataDM.tblOrdersProductID.AsInteger);
end;

procedure TColumnsSimpleDemoMainForm.tvOrdersCarInfoPropertiesCloseUp(Sender: TObject);
begin
  ColumnsSimpleDemoDataDM.tblOrders.Edit;
  ColumnsSimpleDemoDataDM.tblOrdersProductID.Value := dmGridCars.mdModelsID.Value;
end;

procedure TColumnsSimpleDemoMainForm.GridFocusedViewChanged(
  Sender: TcxCustomGrid; APrevFocusedView, AFocusedView: TcxCustomGridView);
begin
  UpdateMenu;
end;

procedure TColumnsSimpleDemoMainForm.miEditButtonsClick(Sender: TObject);
begin
  View.OptionsView.ShowEditButtons := TcxGridShowEditButtons((Sender as TComponent).Tag);
  UpdateMenu;
end;

procedure TColumnsSimpleDemoMainForm.miCellSelectClick(Sender: TObject);
begin
  with View.OptionsSelection do
    CellSelect := not CellSelect;
  UpdateMenu;
end;

procedure TColumnsSimpleDemoMainForm.miHideFocusRectClick(Sender: TObject);
begin
  with View.OptionsSelection do
    HideFocusRectOnExit := not HideFocusRectOnExit;
  UpdateMenu;
end;

procedure TColumnsSimpleDemoMainForm.miHideSelectionClick(Sender: TObject);
begin
  with View.OptionsSelection do
    HideSelection := not HideSelection;
  UpdateMenu;  
end;

procedure TColumnsSimpleDemoMainForm.miInvertSelectClick(Sender: TObject);
begin
  with View.OptionsSelection do
    InvertSelect := not InvertSelect;
  UpdateMenu;  
end;

procedure TColumnsSimpleDemoMainForm.miMultiSelectClick(Sender: TObject);
begin
  with View.OptionsSelection do
    MultiSelect := not MultiSelect;
  UpdateMenu;  
end;

end.  
